### PR TITLE
Fix: #1542. Add Content-Digest.

### PR DESCRIPTION
**Pull request by ioggstream**
## This PR

- [x] adds Content-Digest
- [x] adds Want-Content-Digest
- [x] add examples
- [ ] do we need a further rationale on the discussion?

### CODE REVIEW

**Comment by LPardue** [comment]
I think this is a good start. I took a stab at updating the concept overview to better draw out the differences between Digest and Content-Digest - see https://github.com/httpwg/http-extensions/compare/ioggstream-1542...lucas/content-digest-plusplus?expand=1

Think we should wait for WG feedback before landing this PR into main branch though.

**Comment by ioggstream** [comment]
@LPardue integrated your suggestion in the PR, thanks!

**Comment by LPardue** [comment]
Thanks. (And thanks for putting together the bulk of this pr!)

I think the changes to overview set an example of the kinds of changes we probably need to make throughout the document. But for now what's here is indicative enough of the Content-Digest change to let people assess it does solve their issues.

**Comment by ioggstream** [comment]
After we get consensus on the doc's content, I'm even ok to rewrite everything from scratch :DDD

File: draft-ietf-httpbis-digest-headers.md
Comment (mnot) [inline draft-ietf-httpbis-digest-headers.md:331]: So, strictly speaking, this field name is appropriate according to the current terminology in Semantics. 

What triggered my unease was that the `Content-` prefix is used on several headers that _don't_ refer to the actual content of the message (using the current Semantics definition) -- they refer to the _representation_. See <https://github.com/httpwg/http-core/issues/878>.

I think it's safe to ship this as `Content-Digest` as long as it's being super-precise about what it's referring to.

I could also see flipping `Content-Digest`'s semantics with `Digest`, to align with most (but not all) other uses of the `Content-` prefix. I'm not convinced that doing so is necessary, or even preferable; just mentioning it as something to consider.

Comment (ioggstream) [reply draft-ietf-httpbis-digest-headers.md:331]: Good insight Mark, we'll provide an example of Content-Digest vs Digest in the same message.

About flipping, I think that Content-Digest, being similar to Content-md5 seems to me a more intuitive name.

Comment (LPardue) [reply draft-ietf-httpbis-digest-headers.md:331]: Flipping names seems bad to me. Prefer the way we suggested

Comment (b---c) [reply draft-ietf-httpbis-digest-headers.md:331]: Wouldn't flipping the names be problematic with respect to compatiblity with existing implementations of RFC3230's Digest?

File: draft-ietf-httpbis-digest-headers.md
Comment (MikeBishop) [inline draft-ietf-httpbis-digest-headers.md:162]:
```suggestion
some sort of manipulation to be considered a representation or conveys a
partial representation of a resource, such as Range Requests (see Section
14.2 of {{SEMANTICS}}).
```
There are only two items here, so you don't need a comma before "or"; "for example" would need a comma after it as well, and that would be a lot of commas.

File: draft-ietf-httpbis-digest-headers.md
Comment (MikeBishop) [inline draft-ietf-httpbis-digest-headers.md:164]:
```suggestion
To support use-cases where a simple checksum of the content bytes is required,
```

File: draft-ietf-httpbis-digest-headers.md
Comment (MikeBishop) [inline draft-ietf-httpbis-digest-headers.md:165]:
```suggestion
this document introduces the `Content-Digest` request and response field; see
```

File: draft-ietf-httpbis-digest-headers.md
Comment (MikeBishop) [inline draft-ietf-httpbis-digest-headers.md:335]:
```suggestion
A content digest value is computed by applying a digest-algorithm to the actual message content
```

**Review by MikeBishop — APPROVED** [review]
As noted at the interim, this looks like a worthwhile direction.  Minor editorial nits while reading through, but no disagreements with the proposal.

File: draft-ietf-httpbis-digest-headers.md
Comment (jricher) [inline draft-ietf-httpbis-digest-headers.md:333]: Since we're defining a new header here, can we just use structured fields for it entirely? So it'd be a dictionary with the field names as algorithms and the values as binaries.

Comment (LPardue) [reply draft-ietf-httpbis-digest-headers.md:333]: Excellent question. I like the idea but think that it would make it awkward because the algorithms define their encoding. So you'd have to carve out some exceptions and differences, which might catch implementers out.

Comment (ioggstream) [reply draft-ietf-httpbis-digest-headers.md:333]: I wanted to switch to SF, but after a long discussion it was suggested to retain the current grammar,
even because SF do not support uppercase dictionary keys: this is a show stopper wrt being backward compatible.


Moreover  I don't agree wrt using different parsers  for `C-D` and  `D`: it is valuable that the same parser can process both fields.

**Comment by ioggstream** [comment]
@LPardue do we need examples?

**Comment by ioggstream** [comment]
@LPardue should we create a separate example section for Content-Digest or just add some examples in parallel with Digest?

File: draft-ietf-httpbis-digest-headers.md
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:367]:
```suggestion
validation instead of verifying every received content-digest.
```

File: draft-ietf-httpbis-digest-headers.md
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:388]: Given how similar Want-Digest and Want-Content-Digest are, and that they share the `want-digest-value` we can probably editorialize this a bit. But in order not to block this PR, we can do that work in a follow up.

File: draft-ietf-httpbis-digest-headers.md
Comment (LPardue) [inline draft-ietf-httpbis-digest-headers.md:1318]:
```suggestion
This section registers the `Want-Content-Digest` field in the "Hypertext Transfer
```

File: draft-ietf-httpbis-digest-headers.md
Comment (ioggstream) [reply draft-ietf-httpbis-digest-headers.md:388]: filed https://github.com/httpwg/http-extensions/issues/1632

**Review by LPardue — APPROVED** [review]

**Pull request merged**
